### ISSUE TITLE
[REFACT] Introduce new variables & state modular logic for probabilistic models - [merged]

### ISSUE THREAD

**Issue by NicolasGensollen** [aramis-lab/leaspy#219]
In GitLab by @emaheux on May 24, 2023, 16:25

_Merges eti/refactor-variables -> v2_

### What does the code in the MR do ?

This MR introduces a whole new logic regarding handling of model parameters, latent variables (realizations) and linked attributes (old attributes).
The aim of this change is:
1. to have a more modular way of specifying models, so to ease the definition of new models quickly with very few changes (and limited to 1-2 files max unlike previously).
2. to be more efficient in computations of derived variables (old attributes)

This MR is the first step towards complete paradigm change that shall be implemented in follow-up MRs: 
- it drafts the overall architecture, with main classes for variables that should be almost definitive
- temporarily breaking many models & some functionalities (marked as TODO/WIP/TMP)

### Where should the reviewer start ?

1. You may want to check first the abstract components (utils) that are quite independent of the rest:
- `leaspy.utils.weighted_tensor`
- `leaspy.utils.filtered_mapping_proxy`
- `leaspy.utils.functional` (in particular `NamedInputFunction` logic)

2. Then you can look at the whole `leaspy.variables` architecture:
- Specification of variables --> different classes implementing the `VariableInterface`: `Hyperparameter`, `ModelParameter`, `DataVariable`, `PopulationLatentVariable`, `IndividualLatentVariable`, `LinkedVariable` + a few small classes that are useful
- `VariablesDAG`: Directed Acyclic Graph with nodes being some `VariableInterface`
- `State`: lazy container for values of variables described in a DAG
- `StatelessDistributionFamily` (interface to represent stateless distribution families - i.e. no distribution parameters are stored in instance) and `SymbolicDistribution` (specification of a symbolic distribution, for priors of latent variables and observation models)

3. Then you can look at the new `ObservationModel` which is built on top of `SymbolicDistribution` component and shall supplant the `NoiseModel` (note that models may contain multiple `ObservationModel` at some point, e.g. for joint models but the full implementation of this case of out of this MR scope) 

4. Finally you can look at the changes inside the core Leaspy modules to integrate this new architecture:
- in models: AbstractModel > AbstractMultivariateModel > MultivariateModel > UnivariateModel
- in fit algo (perso algo not done for now)
- in samplers (only very minor changes unlike the previous ones)

(Last commit is only CI-related, fixing behavior of CI pipeline when dealing with git branch containing a slash)

### What are the main remaining points to implement

Here is a recap of what was flagged as TODO/WIP/TMP in the code:
1. Observation models: 
  - Translate all legit noise models into new observation models paradigm, and remove the `leaspy.models.noise_models`
  - Design a keyword-based factory for observation models (for most common use-cases) like the one we had for noise models (cf. comment in AbstractModel `TODO? Factory of `ObservationModel` instead?`) + a proper export routine for those (used when requesting `model.to_dict()`)
  - Simulation algo is left broken for now, and shall be adapted when final observation model(s) interface is OK
2. A bit cleaner `model.initialize`?
3. Logistic parallel model
4. Ordinal model extension (as a dedicated child class?):
   - in particular currently masked samplers is not properly handled with new `WeightedTensor` logic
   - and in stateless distributions we only cope for value being a `WeightedTensor` but not any of the distribution parameters which could be needed for the `batched_deltas` case (but <!> to think about compatibility checks between value & parameters weights if doing so!)
5. Remove all `leaspy.io.realizations` sub-package when ready (in particular move the `VariableType` enum elsewhere, or remove it and re-use the `PopulationLatentVariable` vs. `IndividualLatentVariable` dichotomy instead) and some legacy code that was left commented for now
6. Plotting and saving state in `FitOutputManager` (refactoring the plotting part is a broader work per se)
7. After verification that functional tests pass we may drop the commit f823385 (cf. below in tests section) & all snippets that were "quick&dirty" adaptation to pass the tests + `var_kws.setdefault("scale", var.prior.stddev.call(state))` line as well + regenerate functional data?
8. May be clarify the hyperparameters for models: 
  - there are some hyperparameters that are mandatory to fully specify a model (= be able to instantiate its DAG) (e.g. dimension, source_dimension)
  - but also some `Hyperparameter` variables in DAG (like the fixed population "prior" std-dev) 
  - and also some _hyperparameters_ in the initialization of model parameters (e.g. fixed `NOISE_STD/TAU_STD/XI_STD`)
  - --> may be clarify this in the exported/saved model produced in `.to_dict()` (and reloaded and `.load_parameters()`) so not to mix real model parameters with all those kind of _hyperparameters_
9. Missing some unit tests (some unit tests were written inline in abstract components) + adaptation of some existing ones + missing some docstrings...
10. No implementation of jacobian for now in logistic model (in realty we could use the torch automatic differentiation so not to explicit ourselves the jacobian of variables in our DAG)
11. Reverse model: (biomarker value, individual parameters) -> individual age to reach this value
12. (Design decision about model having an internal state modified by algo [current choice] vs. model only storing model parameters and state only instantiated during algos)
13. (Not sure crucial: `TODO: find a way to prevent re-computation of orthonormal basis since it should not have changed (v0_collinear update)` in the center_xi operation)

And a few other pending TODOs that were not addressed here:
- Proper architecture for models as discussed during meetings
- Finalize metrics/loss handling in algo/models
- Extend `Data` / `Dataset` to be able to cope for different data types (e.g. continuous longitudinal values + survival times observations + ...)
- Currently the state is designed to store values of one iteration only but it could be extended to store the values for all iterations if useful (cf. also comment in State: `TODO? ability to fork after several assignments?`)

### How can the code be tested ?

For this MR many models were broken, the only model really implemented is logistic (univariate or not) with gaussian observation model.

All the corresponding functional tests (for the fit step) should be working:

#### FIT (supported models + obs. model only)

```
pytest tests/functional_tests/api/test_api_fit.py::LeaspyFitTest::test_fit_logistic_scalar_noise tests/functional_tests/api/test_api_fit.py::LeaspyFitTest::test_fit_univariate_logistic tests/functional_tests/api/test_api_fit.py::LeaspyFitTest::test_fit_logistic_diag_noise_with_custom_tuning_no_sources tests/functional_tests/api/test_api_fit.py::LeaspyFitTest::test_fit_logistic_diag_noise_mh tests/functional_tests/api/test_api_fit.py::LeaspyFitTest::test_fit_logistic_diag_noise_fast_gibbs tests/functional_tests/api/test_api_fit.py::LeaspyFitTest::test_fit_logistic_diag_noise
```

#### PERSONALIZE (without offending models, nor scipy_minimize with jacobian)

```
pytest tests/functional_tests/api/test_api_personalize.py
```

#### ESTIMATE (only for supported models + obs_models as of now)

```
pytest tests/functional_tests/api/test_api_estimate.py::LeaspyEstimateTest::test_estimate_multivariate tests/functional_tests/api/test_api_estimate.py::LeaspyEstimateTest::test_estimate_univariate
```

#### FIT + PERSONALIZE (simulation turned off)

```
pytest tests/functional_tests/api/test_api.py::LeaspyAPITest::test_usecase_logistic_scalar_noise tests/functional_tests/api/test_api.py::LeaspyAPITest::test_usecase_logistic_diag_noise
```

(Note that a special commit f823385 > TMP: fix order of variables in loops - was introduced to reproduce the exact same order of variables in sampling as of previously in order to pass those functional tests --> once we are sure about the iso-functionality of MR we should drop this commit and should regenerate functional data)

### When is the MR due for? (review deadline)

?

### What issues are linked to the MR ?

Team discussions about the need for general refactoring of model to easily/properly integrate new models.

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on May 24, 2023, 22:16

added 5 commits

<ul><li>4227e5c2 - [FIX] Reset latent variables at the end of MCMC fit algo to be consistent with...</li><li>905ecb40 - [REFACT] Strategy and place when to put_data_variables into state</li><li>2dbf0100 - [FIX] Working scipy_minimize personalization (without jacobian)</li><li>4866ec2a - [FIX] Increase atol in AbstractModel.load_parameters</li><li>1b7a8924 - [TMP / TESTS] Dirty hacks to check functional tests are OK for subset of implemented stuff</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/97/diffs?diff_id=690257891&start_sha=bd91f3aa81508f06ac7249c948dec3d1de27fcaf)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on May 25, 2023, 14:52

added 1 commit

<ul><li>f3cb0ef4 - [FIX] Estimation of values is now working again + compute_mode|mean_traj +...</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/97/diffs?diff_id=691045113&start_sha=1b7a8924c54a77dc83d1250101226f14b7bad9df)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on May 25, 2023, 15:55

Hello,

First thanks for this hudge MR. I have mostly look at point 1,2,3 in a structural point of view (links between the classes, etc...). To help me going through the MR I have created this graph (not finished...) 

![refacto.drawio](https://my-gitlab-bucket.s3.amazonaws.com/938594763/3d9e382f4b8bf2be455c2d0e7c4dfeaaba33d05acb0977a2e8b34d2baf3603bc/refacto.drawio.png)

I have still a lot to read but a few questions/remarks that I have until now:

_utils:_
* What will be FilteredMappingProxy used for ?
* NamedInputFunction is the type of object that linkedvariable will take as an input ?

_distributions:_
* Nothing for now,

_specs:_
* What does `specs` stand for ? Why not `variables` ?
* What will be NamedVariables and LatentVariableInitType used for ?

_state:_
* What will be StateForkType used for ?

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on May 25, 2023, 16:48

Thanks for this nice diagram! Here are some answers:

# utils

- `FilteredMappingProxy` is used to provide an efficient _sub-dictionary_ (dynamically linked to a reference dictionary), it is only used in `VariablesDAG.sorted_variables_by_type[...]` for now, to provide the subset of variables of a certain type (e.g. `IndividualLatentVariable`) among all variables of the DAG, with a read-only dict (= mapping) interface.
- `LinkedVariable` can take as input:
  - normal functions, with keyword-only parameters [e.g. `def model(*, g, reparam_t, v)`]
  - but since, in some cases, it may boring to create such functions (e.g. `torch.exp` have a positional argument, with fixed name), we provide a convenient wrapper class to produce such function-like objects (= having a `__call__` method)

# specs

- `specs` is used for _specifications_ of variables (the nodes of DAG), `variables.variables` would be a bit strange a found
- `NamedVariables` is a smart dictionary holding `VariableInterface`; to understand why _smart_ cf. the docstring; it is not yet a DAG because conceptually it is different (and in particular it may be instantiated while being _incomplete_)
- `LatentVariableInitType` is used here and there to specify the initialization to use for latent variable (it is strongly link to the `StatelessDistributionFamily` interface provided): one can want the latent variable to be initialized to prior mode  (done in MCMC fit for pop. variable, and in MCMC perso for ind. variable), prior mean (not used in the new code, but we used to talk about `init_at_mean` - where as the mode is most sensed - but since we only used Normal prior before we had `mode = mean` anyway), or random prior samples (done in MCMC fit for ind. variable)

# state

- `StateForkType` enumeration type to specify how the "forking" mechanism (= "rervertible cache" after an assignment) should behave (None = no auto forking, `REF` = store reference of values, `COPY` = store copy of values) - by default I used `REF` but I thought that `COPY` might be needed at some point (but may be not in fact since I strive to only do out-of-place computations for realizations)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on May 26, 2023, 10:08

[I did not have time to finish my comment before the meeting I publish the messy version sorry]

Many thanks for your answer, it quite more clear.

__Noise model:__ I am not quite sure to remember what was the limits of noise model that forced you to create observation model...

![refacto2.drawio](https://my-gitlab-bucket.s3.amazonaws.com/938594763/b9715ce3fe769e866afcd0cb4fe0ba7dbcd3b0ffa0c3964f4a34c82dcb4e23b6/refacto2.drawio.png)

t is not yet a weighted vector ?
(Why is there still a univariate model ?)

Algo fit not done, work ? do not work ? Noise model ?

1. __Objective:__ Mainstream analysis must be possible ? __Tests:__ All the test should work on the basic multivariate model ? 
2. __Objective:__ Improve code structure,  __Tests:__ same as above
3. __Objective:__ Analysis with parallel must be possible [DOES SOMEONE USE THAT?] __Tests:__ test on parallel
4. __Objective:__ Analysis with ordinal model must be possible __Tests:__ test on ordinal
5. __Objective:__ Code structure [WHY NOT POSSIBLE BEFORE ?] __Tests:__ same
6. __Objective:__ Plotting [NEEDED FOR ANALYSIS CANNOT USE BEFORE ?] __Tests:__ same [?]
7. __Objective:__ [?????] __Tests:__ same [?]
8. __Objective:__ Hyperparameters [?????] __Tests:__ same [?]
9. __Objective:__ Improve package testing [OR RESET PAST LEVEL?] __Tests:__ same [?]
10. __Objective:__ [JACOBIAN FOR PERSO] __Tests:__ same [?]
11. __Objective:__ [?] __Tests:__ same [?]
12. __Objective:__ [?] __Tests:__ same [?]
13. __Objective:__ [?] __Tests:__ same [?]

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on May 26, 2023, 13:45

marked this merge request as **ready**

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on May 26, 2023, 13:45

changed target branch from `dev` to `v2`

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on May 26, 2023, 13:47

I'm resolving thread since answers were provided this morning orally during meeting.

I also changed target branch to new *v2* branch (forking from dev). So I think it'd ready to be merged in it as discussed this morning!

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on May 26, 2023, 13:47

resolved all threads

**Comment by NicolasGensollen** [comment]
approved this merge request

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on May 26, 2023, 14:39

approved this merge request

**Comment by NicolasGensollen** [comment]
As discussed this morning, let's merge this in a `v2` branch and iterate there to restore the tests

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on May 26, 2023, 14:45

mentioned in commit 2b18e198ed795fdea4251f1981a58a897a8e919f
